File: typescript/src/n0104_maximum_depth_of_binary_tree.ts
import type { TreeNode } from './utils/tree';

export function maxDepth(root: TreeNode | null): number {
  if (!root) {
    return 0;
  }
  return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
}

export function maxDepthIterative(root: TreeNode | null): number {
  if (!root) {
    return 0;
  }
  const stack = [{ node: root, depth: 1 }];
  let maxDepth = 1;
  while (stack.length > 0) {
    const { node, depth } = stack.pop()!;
    maxDepth = Math.max(maxDepth, depth);
    if (node.left) {
      stack.push({ node: node.left, depth: depth + 1 });
    }
    if (node.right) {
      stack.push({ node: node.right, depth: depth + 1 });
    }
  }
  return maxDepth;
}

File: rust/src/n0104_maximum_depth_of_binary_tree.rs
use std::{cell::RefCell, rc::Rc};

use crate::utils::tree::TreeNode;

pub fn max_depth(root: Option<Rc<RefCell<TreeNode>>>) -> usize {
    match root {
        Some(node) => {
            let left = node.borrow().left.clone();
            let right = node.borrow().right.clone();
            max_depth(left).max(max_depth(right)) + 1
        }
        None => 0,
    }
}

pub fn max_depth_iterative(root: Option<Rc<RefCell<TreeNode>>>) -> usize {
    let mut max_depth = 0;
    if let Some(r) = root {
        let mut stack = vec![(r, 1)];
        while let Some((node, depth)) = stack.pop() {
            max_depth = max_depth.max(depth);
            if let Some(l) = node.borrow().left.clone() {
                stack.push((l, depth + 1));
            }
            if let Some(r) = node.borrow().right.clone() {
                stack.push((r, depth + 1));
            }
        }
    }
    max_depth
}
